filecoin-proofs/src/bin/phase2.rs

```diff
@@ -998,6 +1008,213 @@ fn verify_contribution(path_before: &str, path_after: &str, participant_contrib:
     );
 }
 
+// Non-raw only.
+pub fn read_g1<R: Read>(mut reader: R) -> io::Result<G1Affine> {
+    let mut affine_bytes = G1Uncompressed::empty();
+    reader.read_exact(affine_bytes.as_mut())?;
+    let affine = affine_bytes
+        .into_affine()
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+
+    if affine.is_zero() {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "deserialized G1Affine is point at infinity",
+        ))
+    } else {
+        Ok(affine)
+    }
+}
+
+// Non-raw only.
+pub fn read_g2<R: Read>(mut reader: R) -> io::Result<G2Affine> {
+    let mut affine_bytes = G2Uncompressed::empty();
+    reader.read_exact(affine_bytes.as_mut())?;
+    let affine = affine_bytes
+        .into_affine()
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+
+    if affine.is_zero() {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "deserialized G2Affine is point at infinity",
+        ))
+    } else {
+        Ok(affine)
+    }
+}
+
+fn seek(file: &mut File, offset: u64) -> io::Result<()> {
+    let pos = file.seek(SeekFrom::Start(offset))?;
+    if pos != offset {
+        Err(io::Error::new(
+            io::ErrorKind::UnexpectedEof,
+            format!("seek stopped early, reached: {}, expected: {}", pos, offset),
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+struct FileInfo {
+    delta_g1_offset: u64,
+    delta_g1: G1Affine,
+    delta_g2: G2Affine,
+    h_len: u64,
+    h_offset: u64,
+    h_first: G1Affine,
+    h_last: G1Affine,
+    l_len: u64,
+    l_first: G1Affine,
+    l_last: G1Affine,
+    cs_hash: [u8; 64],
+    contributions_len_offset: u64,
+    contributions_len: u64,
+}
+
+impl Debug for FileInfo {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_struct("FileInfo")
+            .field("delta_g1_offset", &self.delta_g1_offset)
+            .field("delta_g1", &self.delta_g1)
+            .field("delta_g2", &self.delta_g2)
+            .field("h_len", &self.h_len)
+            .field("h_offset", &self.h_offset)
+            .field("h_first", &self.h_first)
+            .field("h_last", &self.h_last)
+            .field("l_len", &self.l_len)
+            .field("l_first", &self.l_first)
+            .field("l_last", &self.l_last)
+            .field("cs_hash", &hex_string(&self.cs_hash))
+            .field("contributions_len_offset", &self.contributions_len_offset)
+            .field("contributions_len", &self.contributions_len)
+            .finish()
+    }
+}
```

Comment:
Out of curiosity, what's the benefit of manually implementing vs deriving `Debug`?

Reply:
^^^ Yup